### PR TITLE
Feature/aggregate data filters

### PR DESCRIPTION
**Pull request by unclecheese**
Per https://github.com/silverstripe/silverstripe-framework/issues/6493

### CODE REVIEW

**Comment by dhensby** [comment]
Postgres issues. Also don't hardcode IDs into tests because postgres won't like that, either.

Linting issues too.

File: tests/php/ORM/DataListTest.php [hunk 1]
```diff
@@ -1220,6 +1222,107 @@ public function testFilterByNullCase()
...
+    	}
+    }
+
+    public function testAggregateFilterExceptions()
+    {
+		$ex = null;
+		try {
+			$filter = new ExactMatchFilter('Comments.Max( This will not parse! )');
+		} catch(\Exception $e) {
+			$ex = $e;
```
Comment (dhensby) [inline tests/php/ORM/DataListTest.php]: It looks like this assignment is unnecessary as the resulting var is never used.

File: tests/php/ORM/DataListTest.php [hunk 2]
```diff
@@ -1220,6 +1222,107 @@ public function testFilterByNullCase()
...
+		$this->assertInstanceOf(\InvalidArgumentException::class, $e);
+		$this->assertRegExp('/Invalid column/', $e->getMessage());
+    }
+
+    public function testAggregateFilters()
+    {    	
+    	$teams = Team::get()->filter('Comments.Count()', 2);
+    	$team1 = $this->objFromFixture(Team::class, 'team1');
+
+    	$this->assertEquals(1, $teams->count());
```
Comment (dhensby) [inline tests/php/ORM/DataListTest.php]: `$this->assertCount(1, $teams)`

File: tests/php/ORM/DataListTest.php [hunk 3]
```diff
@@ -1220,6 +1222,107 @@ public function testFilterByNullCase()
...
+    	$this->assertEquals($team1->ID, $teams->first()->ID);
+
+    	$teams = Team::get()->filter('Comments.Count()', [1,2]);
+    	
+    	$this->assertEquals(2, $teams->count());
+    	$this->assertEquals([], array_diff([1,2], $teams->column('ID')));
```
Comment (dhensby) [inline tests/php/ORM/DataListTest.php]: Don't hard code the IDs as postgres doesn't reset the auto_increment with each test.

My approach tends to be:

```php
$expectedTeam1 = $this->objFromFixture(Team::class, 'team1');
$expectedTeam2 = $this->objFromFixture(Team::class, 'team2');

$this->assertCount(2, $teams);
$this->assertContains($expectedTeam1->ID, $teams->column('ID'));
$this->assertContains($expectedTeam2->ID, $teams->column('ID'));
```

Comment (unclecheese) [inline tests/php/ORM/DataListTest.php]: Thanks, Dan. I thought about that, but it gets awfully long-winded if you have more than a few records in the list. You need an assertion for every ID. Part of the fun, I guess. Thanks for the tip on Postgres. Did not know that.

**Comment by unclecheese** [comment]
Have revised per your review, @dhensby

File: tests/php/ORM/DataListTest.php (hunk 1 above)
Comment (unclecheese) [inline tests/php/ORM/DataListTest.php]: I've updated this to use the `$ex` var instead.

**Comment by dhensby** [comment]
Still lots of code style errors: https://travis-ci.org/silverstripe/silverstripe-framework/jobs/194647196#L641-L1016

Looks like the `GROUP BY` clause is not correctly quoting the fields (they aren't wrapped in `"`)

File: src/ORM/Filters/SearchFilter.php [hunk 4]
```diff
@@ -244,13 +310,34 @@ public function getDbFormattedValue()
...
+     * @return DataQuery
+     */
+    public function applyAggregate(DataQuery $query, $having)
+    {
+    	$schema = DataObject::getSchema();
+    	$baseTable = $schema->baseDataTable($query->dataClass());
+    	
+    	return $query
+        	->having($having)
+        	->groupby("{$baseTable}.ID");
```
Comment (dhensby) [inline src/ORM/Filters/SearchFilter.php]: This line needs to be: `->groupby("\"{$baseTable}\".\"ID\"");`

**Comment by dhensby** [comment]
almost there!

**Comment by unclecheese** [comment]
All sorted, Dan. Cleaned up the code formatting and added those quotes.

**Comment by dhensby** [comment]
Great, waiting for green :)

**Comment by dhensby** [comment]
got a test failure :(

**Comment by dhensby** [comment]
It's postgres - https://travis-ci.org/silverstripe/silverstripe-framework/jobs/195467893#L595

**Comment by dhensby** [comment]
A few CS issues too

**Comment by unclecheese** [comment]
OK, it finally passed. :)

**Comment by tractorcow** [comment]
I know this is a bit late @unclecheese but can we have some user documentation for this? I tried to look it up and I can't find any anywhere.

**Comment by unclecheese** [comment]
Docs are here https://github.com/silverstripe/silverstripe-framework/blob/4/docs/en/02_Developer_Guides/00_Model/01_Data_Model_and_ORM.md#filtering-by-aggregates

**Pull request merged**